### PR TITLE
new module feature-backup

### PR DESCRIPTION
**Pull request by shrsr**
Creates, Deletes and Queries backups

### CODE REVIEW

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+      host: '{{ mso_hostname }}'
+      username: '{{ mso_username }}'
+      password: '{{ mso_password }}'
+      validate_certs: '{{ mso_validate_certs | default(false) }}'
+      use_ssl: '{{ mso_use_ssl | default(true) }}'
+      use_proxy: '{{ mso_use_proxy | default(true) }}'
+      output_level: '{{ mso_output_level | default("info") }}'
+
+# Clean Env for this run
+- name: Remove ansibleBackup1
```
[the remark this answers is not in the document]
Comment (shrsr) [reply tests/integration/targets/mso_backup/tasks/main.yml:24]: Used camel case here because we can't us special chars like '_' for backup names

File: plugins/modules/mso_backup.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        if mso.existing:
+            data = mso.existing
+            mso.existing = []
+            for backup_info in data:
+                backup_id = backup_info.get('id')
+                if backup == backup_info.get('name').split('_')[0]:
+                    path = 'backups/backupRecords/{id}'.format(id=backup_id)
+                    mso.existing.append(backup_info)
+            if len(mso.existing) < 1:
+                mso.existing = mso.get_obj(path, name=backup)
```
[the remark this answers is not in the document]
Comment (shrsr) [reply plugins/modules/mso_backup.py:161]: If the length of mso.existing < 1 it feeds the info to the module that if a backup is not present and we use ansible's absent, then the result is not changed. It's as good as setting mso.exisitng to empty. I can change it to mso.exisitng = {}

File: plugins/modules/mso_backup.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        mso.previous = mso.existing
+        if mso.existing:
+            if module.check_mode:
+                mso.existing = {}
+            else:
+                mso.existing = mso.request(path, method='DELETE')
+        mso.exit_json()
+
+    path = 'backups'
+    mso.existing = mso.get_obj(path, name=backup)
```
[the remark this answers is not in the document]
Comment (shrsr) [reply plugins/modules/mso_backup.py:176]: We can change it to mso.existing = {} 
We need this to be able to do a clean POST operation every time because despite having the same name for backup, we create a new backup from scratch. We wouldn't need it if we had PUT for the backup to be changed.

File: tests/integration/targets/mso_backup/tasks/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lhercot) [reply tests/integration/targets/mso_backup/tasks/main.yml:24]: Ok for the backup name but then use snake case for the registered variable

[the remark this answers is not in the document]
Comment (shrsr) [reply tests/integration/targets/mso_backup/tasks/main.yml:24]: It complains that is..

File: plugins/modules/mso_backup.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    type: str
+    aliases: [ name ]
+  remote_location:
+    description:
+    - The remote location's name for the backup to be stored
+    type: str
+  remote_path:
+   description:
+    - This path is relative to the remote location.
+    - There's no need to specify '/' for this path.
```
Comment (lhercot) [inline plugins/modules/mso_backup.py:41]: I would rephrase with "A '/' is automatically added between the remote location folder and this path."
I would also add 'This folder structure should already exist on the remote location'.

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+  register: query_ansibleBackup_for_delete
+
+- name: Remove backups with starting with name ansibleBackup
+  mso_backup:
+    <<: *mso_info 
+    backup:  '{{ item.name }}'
+    state: absent
+  loop: '{{ query_ansibleBackup_for_delete.current | selectattr("name", "match", "^ansibleBackup.*") | list  }}'
+
+- name: Query all backups  post
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:36]: Now that we know it work, we can remove this query post delete.
